kcidb/test_oo.py

```diff
@@ -203,3 +204,250 @@ def test_two_object(self):
             '>checkout#>build#>test#',
             {"test", "build", "checkout", "revision"}, 2
         )
+
+
+class KCIDBTraversingTestCase(kcidb.unittest.TestCase):
+    """Test case for traversing objects."""
+
+    # Conform to unittest conventions, pylint: disable=invalid-name
+
+    def query_str(self, pattern_string):
+        """Run OO client query with a parsed string."""
+        return self.client.query(
+            kcidb.orm.Pattern.parse(pattern_string)
+        )
+
+    def assertContains(self, container, contents):
+        """
+        Check that a container has the specified contents.
+
+        Args:
+            container:  The container to check the contents of.
+            contents:   The contents pattern to check against. The pattern
+                        matches either a list of certain number of objects
+                        of particular type, or a dictionary with string keys
+                        and values matching specified patterns.
+
+                        A list pattern is a tuple containing the type of each
+                        contained instance, and the number of those instances.
+
+                        A dictionary pattern is a dictionary containing string
+                        keys for the matched dictionary to have, with values
+                        being patterns for the corresponding values in the
+                        matched dictionary.
+        """
+        assert isinstance(container, (list, dict))
+        assert isinstance(contents, (tuple, dict))
+
+        if isinstance(contents, tuple):
+            self.assertEqual(
+                len(contents), 2,
+                msg=f"Invalid number of element in contents"
+            )
+            self.assertEqual(
+                len(container), contents[1],
+                msg=f"Invalid number of {contents[0]} instances"
+            )
+            for i, obj in enumerate(container):
+                self.assertIsInstance(
+                    obj, contents[0],
+                    msg=f"Invalid type of {contents[0]} #{i}")
+
+        elif isinstance(contents, dict):
+            key = contents.keys() ^ container.keys()
+            self.assertFalse(key, msg=f"{key} found.")
```

Comment:
This is a little too terse. Could you make a separate assertion for missing keys, and another for extra keys, each with a clear message to avoid head scratching in case of failure?

Otherwise looks great, thank you :blush: